src/topology_cacher.py

```diff
@@ -0,0 +1,567 @@
+      ],
+      "tags": []
+    },
+    "AGLT2-squid-2": {
+      "fqdn": "sl-um-es3.slateci.io",
+      "group_name": "AGLT2",
+      "name": "AGLT2-squid-2",
+      "service_ids": [
+        "138"
+      ],
+      "tags": []
+    },
+    ...
+  }
+}
+```
+There's some redundancy in the information (e.g. fqdn is included in the
+`resources_by_fqdn` entries) but having a consistent entry format makes
+things easier to read (and was easier to implement).
+
+service_ids are numeric but if we're committed to never changing service
+names then I could make them text instead.
+
+"""
+from argparse import ArgumentParser
+from collections import namedtuple
+import json
+import logging
+import os
+import sys
+import xml.etree.ElementTree as ET
+
+
+from urllib.request import urlopen
+
+from typing import Optional
+
+
+TOPOLOGY = "https://topology.opensciencegrid.org"
+
+
+log = logging.getLogger(__name__)
+
+
+class DataError(Exception):
+    pass
+
+
+class ResourceInfo(namedtuple("ResourceInfo", "group_name name fqdn service_ids tags")):
+    SERVICE_ID_CE = "1"
+    SERVICE_ID_SCHEDD = "109"
+
+    def is_ce(self):
+        return self.SERVICE_ID_CE in self.service_ids
+
+    def is_schedd(self):
+        return self.SERVICE_ID_SCHEDD in self.service_ids
+
+
+class TopologyData:
+    def __init__(
+        self,
+        topology_base=TOPOLOGY,
+    ):
+        if "://" not in topology_base:
+            topology_base = "https://" + topology_base
+        self.topology_base = topology_base
+        self.resinfo_table = []
+        self.grouped_resinfo = {}
+        self.resinfo_by_name = {}
+        self.resinfo_by_fqdn = {}
+        self.projects = self.get_projects()
+        self.resources = self.get_resources()
+
+    def get_projects(self) -> ET.Element:
+        return self._get_data("/miscproject/xml")
+
+    def get_resources(self) -> ET.Element:
+        data = self._get_data("/rgsummary/xml")
+        self.resinfo_table = []
+        self.grouped_resinfo = {}
+        self.resinfo_by_name = {}
+        self.resinfo_by_fqdn = {}
+
+        #
+        # Build tables and indices for easier lookup
+        #
+        for eResourceGroup in data.findall("./ResourceGroup"):
+            group_name = safe_element_text(eResourceGroup.find("./GroupName"))
+            if not group_name:
+                log.warning(
+                    "Skipping malformed ResourceGroup: %s", elem2str(eResourceGroup)
+                )
+                continue
+            self.grouped_resinfo[group_name] = []
+
+            for eResource in eResourceGroup.findall("./Resources/Resource"):
+                resource_name = safe_element_text(eResource.find("./Name"))
+                fqdn = safe_element_text(eResource.find("./FQDN"))
+                service_ids = list(
+                    filter(
+                        None,
+                        map(safe_element_text, eResource.findall("./Services/Service/ID"))
+                    )
+                )
```

Comment:
erp, sorry i had meant for you to use findall_nonempty here also:

```suggestion
                service_ids = findall_nonempty(eResource, "./Services/Service/ID")
```

... i mean, it is nice, isn't it?